platform/BufferedSerial.h

```diff
@@ -0,0 +1,94 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_BUFFEREDSERIAL_H
+#define MBED_BUFFEREDSERIAL_H
+
+#include "platform/platform.h"
+
+#if DEVICE_SERIAL
+
+#include "FileHandle.h"
+#include "SerialBase.h"
+#include "PlatformMutex.h"
+#include "serial_api.h"
+#include "CircularBuffer.h"
+
+namespace mbed {
+class BufferedSerial : private SerialBase, public FileHandle {
```

Comment:
@geky Yes, true. Similar codes with different names living at disparate locations is a nightmare from maintenance point of view. But, I think we don't have another option within the time-frame.
When you say drop the file handle stuff from AT parser and BufferedSerial, then why don't you simply say that drop this PR :) ? This PR loses all it stands for if we go the way you prescribed. 
Regarding updating the external repos first suggestion, I'm not sure what will that achieve. APIs will still be broken for whatever users and eventually those would move in we all want these to be part of mbed-os tree, don't we ? .. I totally understand that its critical for us to bring the most solid form of APIs to get in, I'm all open for discussion, suggestion and brainstorming. But aren't we running out of time here ? We are not in requirement planning stage or architecture design phase righ now. Anyway, its a good lesson for us and we have got a lot of pointers regarding future endeavours. I hope those will pass smoother than this :)